Waku v2: History response pagination specs (#218)
vacp2p/rfc: specs/waku/v2/waku-store.md

---
title: Waku
version: 2.0.0-beta2
status: Draft
authors: Oskar Thorén <oskar@status.im>, Dean Eigenmann <dean@status.im>, Sanaz Taheri <sanaz@status.im>
---

# Table of Contents

> - [Table of Contents](#table-of-contents)

Reviewer: ```suggestion
```

- [Abstract](#abstract)
- [Wire Specification](#wire-specification)
  - [Protobuf](#protobuf)
    - [Index](#index)
    - [PagingInfo](#paginginfo)
    - [HistoryQuery](#historyquery)
    - [HistoryResponse](#historyresponse)
- [Changelog](#changelog)
- [Copyright](#copyright)

# Abstract

> `WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes. It also supports pagination for more efficient querying of historical messages where a querying node can set the number of `WakuMessage`s to return and a cursor to start from. The queried node MUST then use this information to return messages in the queried range. The query response also embodies an updated cursor which enables the querying node resume its subsequent requests in an iterative manner.

Reviewer: I'd either move or remove:

> The queried node MUST then use this information to return messages in the queried range. The query response also embodies an updated cursor which enables the querying node resume its subsequent requests in an iterative manner.

It is a bit too detailed/spec-like for an abstract/executive summary with language such as MUST etc.

As a reader skimming an abstract, I only want to know roughly what functionality is provided and why it exists, not know how to use it or exact guarantees for API calls. The shorter the better!

Author: Sure, addressed

**Protocol identifier***: `/vac/waku/store/2.0.0-beta2`

# Wire Specification

> Peers communicate with each other using a request/response API. The messages sent are Protobuf RPC messages. The followings are the specifications of the Protobuf messages. 

Reviewer 2: ```suggestion
Peers communicate with each other using a request / response API. The messages sent are Protobuf RPC messages. The followings are the specifications of the Protobuf messages. 
```

Not that important but I would readd the spaces.

## Protobuf

```protobuf
message Index {

>   bytes digest=1
>   float receivedTime=2
> }
> 
> message PagingInfo {
>   int64 pageSize=1 
>   Index cursor=2

Reviewer 2: ```suggestion
  bytes digest = 1;
  float receivedTime = 2;
}

message PagingInfo {
  int64 pageSize = 1;
  Index cursor = 2;
```

  enum Direction {
    FORWARD = 0;
    BACKWARD = 1;
  }
  Direction direction = 3;
}

message HistoryQuery {
  repeated string topics = 2;

>   optional PagingInfo pagingInfo = 3 // used for pagination
> 
> }

Reviewer 2: ```suggestion
  optional PagingInfo pagingInfo = 3; // used for pagination
}
```

message HistoryResponse {
  repeated WakuMessage messages = 2;

>   optional PagingInfo pagingInfo = 3 // used for pagination

Reviewer 2: ```suggestion
  optional PagingInfo pagingInfo = 3; // used for pagination
```

title: Waku

> version: 2.0.0-beta1

Reviewer: Bump to beta2

Reviewer: And add changelog

status: Draft
authors: Oskar Thorén <oskar@status.im>, Dean Eigenmann <dean@status.im>, Sanaz Taheri <sanaz@status.im>
---

# Table of Contents

- [Abstract](#abstract)

>   * [Waku Store Protocol](#Waku-Store-Protocol)

Reviewer: Usually this is autogenerated with markdown-toc and result would be lower-case-links. Easier to avoid mistakes and do less with this tool

  * [Pagination](#Pagination)
- [Wire Specification](#wire-specification)
  * [Protobuf](#protobuf)
- [Changelog](#changelog)
- [Copyright](#copyright)

# Abstract

> ## Waku Store Protocol 

Reviewer: ```suggestion
```
I think this still makes sense as an abstract, where pagination is one aspect of the store protocol. Otherwise they sound disjointed to me

> `WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes.
> 

Reviewer: ```suggestion
`WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes. It also support pagination for more efficient querying of historical messages.

```

> ## Pagination

Reviewer: I'd probably pull this down to under wire specification or as part of it. So it'd be something like:

Abstract
Motivation (currently missing and probably out of scope for PR)
Wire spec
 - details about fields
   - possibly on two levels if needed

E.g. something like this https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-relay.md#rpc

Reviewer: Right now it seems like it is a lot of duplication, I might be wrong and we need a separate section (e.g. before or after wire spec), but I think it can be done as part of it.

Reviewer: If we want to add motivation, it'd be something about nodes being offline a lot and wanting to get historical messages, etc. E.g. how to address subset of https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-v2.md#motivation-and-goals

Author: Sure, I'll include the motivation, perhaps it should cover motivation for the store protocol as well as the pagination, right?

Author: In the abstract section, my aim was to provide an overview of the pagination process and to give an insight into 
why the protobuf messages that are provided in the wire spec are even needed. I thought it would be easier to understand the whole spec this way. However, I agree that the current writeup has duplications, and the wire specs or the abstract must be shorten.

Reviewer: I see. To me, the abstract should be quite concise and max 200 words (probably less in this case because it isn't a very complex protocol) without too much technical detail. If that can be achieved without detailing implementation details it seems fine to have a sentence or three re pagination in there.

Reviewer: Kind of like an executive summary to get a high level overview of what the store protocol is and why one should care

Author: > I'd probably pull this down to under wire specification or as part of it. So it'd be something like:
> 
> Abstract
> Motivation (currently missing and probably out of scope for PR)

I will open a PR for the motivation

> ## Pagination
> Nodes supporting store protocol MAY support Pagination. When a node supports pagination, a requester node can set the number of waku messages to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.

Reviewer: ```suggestion
## Pagination

Nodes supporting the store protocol MAY support pagination. When a node supports pagination, a requester node can set the number of `WakuMessage`s to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
```

> To perform pagination, each `WakuMessage` stored at a waku store node is associated with a unique `Index`.  Once a `WakuMessage` is received by a node, it will generate the `Index` of the  `WakuMessage` and then saves the (`Index`, `WakuMessage`) together in its local storage. An `Index` encapsulates fields useful for paging, these fields are a `digest` and a `receivedTime`. The `receivedTime` indicates the time at which the `WakuMessage` is received by the store node and the `digest` is the hash of the received `WakuMessage`.  

Reviewer: ```suggestion
To perform pagination, each `WakuMessage` stored at a Waku node running the store protocol is associated with a unique `Index`. Once a `WakuMessage` is received by a node, it will generate the `Index` of the  `WakuMessage` and then saves the (`Index`, `WakuMessage`) together in its local storage. An `Index` encapsulates fields useful for paging, these fields are a `digest` and a `receivedTime`. The `receivedTime` indicates the time at which the `WakuMessage` is received by the store node and the `digest` is the hash of the received `WakuMessage`.  
```

When a `HistoryQuery` arrives, it may contain an extra field `pagingInfo` that carries the information required for paging. The `pagingInfo` contains the `pageSize`, `cursor` and the `direction` as defined below:

**pageSize**

- A positive integer indicating the number of `WakuMessage`s to be included in the `HistoryResponse`
- If the `pageSize` is omitted then a “default page size” shall be used
- There is an upper limit for the page size
- If the number of remaining `WakuMessage`s is less than the `pageSize`, then the `pageSize` is adjusted based on the remaining messages

**Cursor**

-  Holds the `Index` of a waku message

**Direction**

> - Is a boolean value
>      - True means forward pagination that is the `messages` field of the `HistoryResponse` contains `pageSize` many (or less) waku messages whose `Index` are larger than the given `cursor`
>      - False means backward pagination, that is the `messages` field of the `HistoryResponse` contains  `pageSize` many (or less) waku messages whose `Index` are smaller than the given cursor

Reviewer: ```suggestion
- Is a boolean value
     - True means forward pagination that is the `messages` field of the `HistoryResponse` contains `pageSize` many (or less) waku messages whose `Index` are larger than the given `cursor`
     - False means backward pagination, that is the `messages` field of the `HistoryResponse` contains  `pageSize` many (or less) waku messages whose `Index` are smaller than the given cursor
```

I think this can  probably be written more concisely in one or two sentences

> Given a valid  `pagingInfo`, the waku store node performs as follows:

Reviewer: This seems like a HOW vs WHAT. I think as long as we specfiy WHAT is returned the HOW will be an implementation detail, or be immediately apparent (e..g sorting, etc)

Author: I am wondering whether the execution details, like this one, will be ever included in any doc? for example, in the .md files of the nim-waku repo? Or explaining these details is outside of the scope (details of the pagination may seem simple, but I am asking for general cases)?

Reviewer: I think either in-line as part of code comments, e.g. like https://github.com/status-im/nim-waku/blob/master/waku/protocol/v1/waku_protocol.nim#L11-L38 or if it is more of an API thing, with markdown files in doc folder like https://github.com/status-im/nim-waku/blob/master/docs/api/v2/node.md

That said, docs are currently lacking and open to ideas for how to best improve them!

- sorts the list of `WakuMessage`s based on their `Index`
- finds the `WakuMessage` whose `Index` is the same as `cursor` 
- Extracts `pageSize` many waku messages whose `Index` exceed/precede the `cursor` (depending on the `direction` being true/false)
- the resultant list  is sent in the  `messages` field of the `HistoryResponse`

> The `HistoryResponse` will also contain an updated `pagingInfo` which helps the requester to resume its further requests; the `pageSize` indicates the number of returned waku messages (i.e., the number of messages included in `messages` field of `HistoryResponse`), the `direction` is the same direction as in the corresponding `HistoryQuery`. The `cursor` holds the `Index` of the last/first message in the `HistoryResponse` `messages` (depending on the `direction` being true/false).

Reviewer: For true/false statements, I'd probably write it explicitly with one case, and then write something like "...and vice versa for direction being false" or "...and true and false respectively for foo".

Reviewer: Might want to add a sentence or paragraph on how this allows a querying node to receive all messages in an iterated manner, either starting from the beginning or from the end.

Also, probably belongs further up as a general comment, all of this assumes the node was online to receive messages from relay protocol. There could still be gaps :<

Author: By an online node you meant the responder node is supposed to be online and receive all the messages, right?Then sure, gaps are inevitable, indeed we never attempted this problem (gaps) in the current pagination method, the current design only works for one requester and responder. Please let me know if a separate issue is needed to address the offline (store) nodes?

Reviewer: Yeah, I think just a sentence or so to make it clear to the reader what the assumption is, since it might not be obvious to everyone.

I think there might already be an issue for the general store high availability requirement problem, (@decanus might know), but feel free to create one if you want! Definitely not a part of the scope of the current work though, as you say.

> **Protocol identifier***: `/vac/waku/store/2.0.0-beta1`

Reviewer: Should be in abstract for easy reference I think

title: Waku
version: 2.0.0-beta1
status: Draft
authors: Oskar Thorén <oskar@status.im>, Dean Eigenmann <dean@status.im>, Sanaz Taheri <sanaz@status.im>
---

# Table of Contents

- [Table of Contents](#table-of-contents)
- [Abstract](#abstract)
  - [Waku Store Protocol](#waku-store-protocol)
  - [Pagination](#pagination)
- [Wire Specification](#wire-specification)
  - [Protobuf](#protobuf)
    - [Index](#index)
    - [PagingInfo](#paginginfo)
    - [HistoryQuery](#historyquery)
    - [HistoryResponse](#historyresponse)
    - [HistoryRPC](#historyrpc)
- [Changelog](#changelog)
- [Copyright](#copyright)

# Abstract

`WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes. It also support pagination for more efficient querying of historical messages.

## Pagination

> Nodes supporting the store protocol MAY support pagination. When a node supports pagination, a requester node can set the number of `WakuMessage`s to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.

Reviewer 2: ```suggestion
Nodes supporting the store protocol MAY support pagination. When a node supports pagination, a querying node can set the number of `WakuMessage`s to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
```

> To perform pagination, each `WakuMessage` stored at a Waku node running the store protocol is associated with a unique `Index`. Once a `WakuMessage` is received by a node, it will generate the `Index` of the  `WakuMessage` and then saves the (`Index`, `WakuMessage`) together in its local storage. An `Index` encapsulates fields useful for paging, these fields are a `digest` and a `receivedTime`. The `receivedTime` indicates the time at which the `WakuMessage` is received by the store node and the `digest` is the hash of the received `WakuMessage`.  

Reviewer 2: So I think that the `Index` is an implementation detail and therefore not required to be in the spec, wdyt @oskarth?

Reviewer: Index is a protobuf field so it should be explained to same extent (type, basic semantics). Agree it can be more minimal though (see duplication comment)

When a `HistoryQuery` arrives, it may contain an extra field `pagingInfo` that carries the information required for paging. The `pagingInfo` contains the `pageSize`, `cursor` and the `direction` as defined below:

**pageSize**

- A positive integer indicating the number of `WakuMessage`s to be included in the `HistoryResponse`
- If the `pageSize` is omitted then a “default page size” shall be used
- There is an upper limit for the page size
- If the number of remaining `WakuMessage`s is less than the `pageSize`, then the `pageSize` is adjusted based on the remaining messages

**Cursor**

-  Holds the `Index` of a waku message

**Direction**

> - Is a boolean value

Reviewer 2: @staheri14 to limit the need of all this explanation, it probably makes it easier to just make direction an enum

Author: done

}
```

### Index

> To perform pagination, each `WakuMessage` stored at a node running the store protocol is associated with a unique `Index` that encapsulates the following parts. 

Reviewer: ```suggestion
### Index

To perform pagination, each `WakuMessage` stored at a node running the store protocol is associated with a unique `Index` that encapsulates the following parts. 
```

- `digest`:  a sequence of bytes representing the hash of a `WakuMessage`.
- `receivedTime`: the UNIX time at which the waku message is received by the node running the store protocol.

### PagingInfo

> `PagingInfo` holds the information required for pagination. It consists of the following components.

Reviewer: ```suggestion
### PagingInfo

`PagingInfo` holds the information required for pagination. It consists of the following components.
```

- `pageSize`: A positive integer indicating the number of queried `WakuMessage`s in a `HistoryQuery` (or retrieved  `WakuMessage`s in a `HistoryResponse`).
- `cursor`: holds the `Index` of a `WakuMessage`.
- `direction`: indicates the direction of paging which can be either `FORWARD` or `BACKWARD`.

### HistoryQuery

RPC call to query historical messages.

- The `topics` field MUST indicate the list of topics to query.

> - `PagingInfo` holds the information required for pagination.  Its `pageSize` field indicates the number of  `WakuMessage`s to be included in the corresponding `HistoryResponse`. If the `pageSize` is omitted or exceeds an upper limit then a “default page size” shall be used instead. In the forward pagination request, the `messages` field of the `HistoryResponse` shall contain `pageSize` many waku messages whose `Index` are larger than the given `cursor` (and vise versa for the backward pagination). Note that the `cursor` of a `HistoryQuery` may be empty (e.g., for the initial query), as such, the store node, depending on whether the  `direction` is `BACKWARD` or `FORWARD` returns the last or the first `pageSize` waku messages, respectively.

Reviewer 2: ```suggestion
- `PagingInfo` holds the information required for pagination.  Its `pageSize` field indicates the number of  `WakuMessage`s to be included in the corresponding `HistoryResponse`. If the `pageSize` is omitted or exceeds an upper limit then a "default page size" shall be used instead. In the forward pagination request, the `messages` field of the `HistoryResponse` shall contain `pageSize` many waku messages whose `Index` are larger than the given `cursor` (and vise versa for the backward pagination). Note that the `cursor` of a `HistoryQuery` may be empty (e.g., for the initial query), as such, the store node, depending on whether the  `direction` is `BACKWARD` or `FORWARD` returns the last or the first `pageSize` waku messages, respectively.
```

### HistoryResponse

RPC call to respond to a HistoryQuery call.
- The `messages` field MUST contain the messages found, these are [`WakuMessage`] types as defined in the corresponding [specification](./waku-message.md).
- `PagingInfo`  holds the paging information based on which the querying node can resume its further history queries. The `pageSize` indicates the number of returned waku messages (i.e., the number of messages included in the `messages` field of `HistoryResponse`). The `direction` is the same direction as in the corresponding `HistoryQuery`. In the forward pagination, the `cursor` holds the `Index` of the last message in the `HistoryResponse` `messages` (and the first message in the backward paging). The requester shall embed the returned  `cursor` inside its next `HistoryQuery` to retrieve the next page of the waku messages.  The  `cursor` obtained from one node SHOULD NOT be used in a request to another node because the result MAY be different.

# Changelog

2.0.0-beta2

> Initial draft version. Released [2020-10-21](https://github.com/vacp2p/specs/commit/75b4c39e7945eb71ad3f9a0a62b99cff5dac42cf)

Reviewer: The way we do releases is:

1) People approve PR
2) Squash commits in PR
3) Add one more commit to PR with commit info to previously squashed PR
4) Merge rebase

This means the PR results in two commits on master, with the commit hash of a release leading to all the changes done in that release (i.e. this PR).

---

Also the 2.0.0-beta1 should still exist, just appended to in reverse chronological order. Then the beta2 will have a changelog with something like:

- Added pagination support
- Clarified X

Author: Thanks for the clarification :)

}
```

### Index

To perform pagination, each `WakuMessage` stored at a node running the store protocol is associated with a unique `Index` that encapsulates the following parts. 
- `digest`:  a sequence of bytes representing the hash of a `WakuMessage`.
- `receivedTime`: the UNIX time at which the waku message is received by the node running the store protocol.

### PagingInfo

`PagingInfo` holds the information required for pagination. It consists of the following components.
- `pageSize`: A positive integer indicating the number of queried `WakuMessage`s in a `HistoryQuery` (or retrieved  `WakuMessage`s in a `HistoryResponse`).
- `cursor`: holds the `Index` of a `WakuMessage`.
- `direction`: indicates the direction of paging which can be either `FORWARD` or `BACKWARD`.

### HistoryQuery

RPC call to query historical messages.

- The `topics` field MUST indicate the list of topics to query.

> - `PagingInfo` holds the information required for pagination.  Its `pageSize` field indicates the number of  `WakuMessage`s to be included in the corresponding `HistoryResponse`. If the `pageSize` is omitted or exceeds an upper limit then a "default page size" shall be used instead. In the forward pagination request, the `messages` field of the `HistoryResponse` shall contain `pageSize` many waku messages whose `Index` are larger than the given `cursor` (and vise versa for the backward pagination). Note that the `cursor` of a `HistoryQuery` may be empty (e.g., for the initial query), as such, the store node, depending on whether the  `direction` is `BACKWARD` or `FORWARD` returns the last or the first `pageSize` waku messages, respectively.

Reviewer 3: This reads a bit difficult and is probably grammatically not fully correct.
Perhaps this is better? Not sure.
```suggestion
- `PagingInfo` holds the information required for pagination.  Its `pageSize` field indicates the number of  `WakuMessage`s to be included in the corresponding `HistoryResponse`. If the `pageSize` is omitted or exceeds an upper limit then a "default page size" shall be used instead. In the forward pagination request, the `messages` field of the `HistoryResponse` shall contain at maximum the `pageSize` amount of waku messages whose `Index` values are larger than the given `cursor` (and vise versa for the backward pagination). Note that the `cursor` of a `HistoryQuery` may be empty (e.g., for the initial query), as such, the store node, depending on whether the  `direction` is `BACKWARD` or `FORWARD` returns the last or the first `pageSize` waku messages, respectively.
```

### HistoryResponse

RPC call to respond to a HistoryQuery call.
- The `messages` field MUST contain the messages found, these are [`WakuMessage`] types as defined in the corresponding [specification](./waku-message.md).
- `PagingInfo`  holds the paging information based on which the querying node can resume its further history queries. The `pageSize` indicates the number of returned waku messages (i.e., the number of messages included in the `messages` field of `HistoryResponse`). The `direction` is the same direction as in the corresponding `HistoryQuery`. In the forward pagination, the `cursor` holds the `Index` of the last message in the `HistoryResponse` `messages` (and the first message in the backward paging). The requester shall embed the returned  `cursor` inside its next `HistoryQuery` to retrieve the next page of the waku messages.  The  `cursor` obtained from one node SHOULD NOT be used in a request to another node because the result MAY be different.

# Changelog

### 2.0.0-beta2 

> Released [2020-10-27]()

Reviewer: Squash to one commit and add release commit, then we can merge!

}
```

### Index
`Index` has two parts:
- `digest`: holds the hash value as a sequence of bytes
- `receivedTime`: the UNIX time at which the message is received by the waku store node

> ### PagingInfo
> `PagingInfo` has three components defined below.

Reviewer: ```suggestion

### PagingInfo
`PagingInfo` has three components defined below.
```

Ditto above

- `pageSize`: A positive integer indicating the number of waku messages that the client wants to see in her response
- `cursor`: holds the `Index` of a `WakuMessage`
- `direction`: It is a boolean value indicating the direction of paging, `true` stands for forward paging and `false` means backward

### HistoryQuery

RPC call to query historical messages.

- The `topics` field MUST indicate the list of topics to query.
- `PagingInfo` holds the information required for pagination.

### HistoryResponse

RPC call to respond to a HistoryQuery call.
- The `messages` field MUST contain the messages found, these are [`WakuMessage`] types as defined in the corresponding [specification](./waku-message.md).
- `PagingInfo`  holds paging information based on which the requester can resume its further history queries. 

### HistoryRPC

> A node MUST send all History messages (`HistoryQuery`, `HistoryResponse`) wrapped inside a
> `HistoryRPC`.

Reviewer: ```suggestion
A node MUST send all History messages (`HistoryQuery`, `HistoryResponse`) wrapped inside a `HistoryRPC`.
```

> - The `request_id` MUST be a uniquely generated string, the `HistoryResponse` and `HistoryQuery` `request_id` MUST match. 

Reviewer: Formatting seems off, either - for each, newline or as one paragraph

### HistoryQuery

RPC call to query historical messages.

- The `topics` field MUST indicate the list of topics to query.
- `PagingInfo` holds the information required for pagination.

### HistoryResponse

RPC call to respond to a HistoryQuery call.
- The `messages` field MUST contain the messages found, these are [`WakuMessage`] types as defined in the corresponding [specification](./waku-message.md).
- `PagingInfo`  holds paging information based on which the requester can resume its further history queries. 

### HistoryRPC

A node MUST send all History messages (`HistoryQuery`, `HistoryResponse`) wrapped inside a `HistoryRPC`.

- The `request_id` MUST be a uniquely generated string, the `HistoryResponse` and `HistoryQuery` `request_id` MUST match. 
The `HistoryResponse` message SHOULD contain any messages found whose `topic` can be found in the `HistoryQuery` `topics` array. Any message whose `topic` does not match MUST NOT be included. Also, the `HistoryResponse` messages SHOULD contain messages that fall into the range specified by the `pagingInfo` field of the `HistoryQuery`.

# Changelog

2.0.0-beta1

> Initial draft version. Released [2020-10-21](https://github.com/vacp2p/specs/commit/75b4c39e7945eb71ad3f9a0a62b99cff5dac42cf)

Reviewer 2: New version

Author: addressed

Reviewer 2: @staheri14 there should still be the old version, don't remove the old one. Just add a new one
